File: frontend/src/effects/GoL/GoL.js
import QuantumCircuit from 'quantum-circuit';
import * as utils from '../../utils/imageUtils.js';
import * as colors from '../../utils/colorUtils.js';

const MAPPING_TYPES = ['hsv', 'hvs', 'shv', 'svh', 'vhs', 'vsh'];

function hsv_to_statevector(hsv_arr, mapping) {
    let hsv = { h: hsv_arr[0], s: hsv_arr[1], v: hsv_arr[2] };
    let phi = hsv[mapping[0]] * 2 * Math.PI;
    let theta = hsv[mapping[1]] * Math.PI;

    let alpha = Math.cos(theta / 2);
    let beta_re = Math.cos(phi) * Math.sin(theta / 2);
    let beta_im = Math.sin(phi) * Math.sin(theta / 2);

    return [{ re: alpha, im: 0 }, { re: beta_re, im: beta_im }, hsv[mapping[2]]];
}

function statevector_to_hsv(state, semiclassical, mapping) {
    let alpha_re = state[0].re;
    let beta_re = state[1].re;
    let beta_im = state[1].im;

    let theta = 2 * Math.acos(Math.max(0, Math.min(1, Math.abs(alpha_re))));
    let phi = 0;
    if (Math.abs(beta_re) > 1e-9 || Math.abs(beta_im) > 1e-9) {
        phi = Math.atan2(beta_im, beta_re);
    }

    let h = phi / (2 * Math.PI);
    if (h < 0) h += 1;
    let v = theta / Math.PI;

    let res = [0, 0, 0];
    let mapIdx = { 'h': 0, 's': 1, 'v': 2 };
    res[mapIdx[mapping[0]]] = h;
    res[mapIdx[mapping[1]]] = Math.max(0, Math.min(1, v));
    res[mapIdx[mapping[2]]] = semiclassical;
    return res;
}

function find_closest_pure_state(rho00, rho11, rho01_re, rho01_im) {
    let a = rho00;
    let b = rho11;
    let c2 = rho01_re * rho01_re + rho01_im * rho01_im;
    let purity = a * a + b * b + 2 * c2;

    let trace = a + b;
    let det = a * b - c2;
    let discriminant = Math.max(0, trace * trace - 4 * det);
    let lambda = (trace + Math.sqrt(discriminant)) / 2;

    if (c2 < 1e-12) {
        if (a > b) return [[{ re: 1, im: 0 }, { re: 0, im: 0 }], purity];
        else return [[{ re: 0, im: 0 }, { re: 1, im: 0 }], purity];
    }

    let x_re = rho01_re, x_im = rho01_im;
    let y_re = lambda - a, y_im = 0;

    let norm = Math.sqrt(x_re * x_re + x_im * x_im + y_re * y_re + y_im * y_im);
    return [[{ re: x_re / norm, im: x_im / norm }, { re: y_re / norm, im: y_im / norm }], purity];
}

function liveliness(nhood_s) {
    let sum = 0;
    for (let i of [0, 1, 2, 3, 5, 6, 7, 8]) {
        sum += nhood_s[i][0];
    }
    return Math.abs(sum);
}

function SCGOL(nhood_s) {
    let a = liveliness(nhood_s);
    let value = nhood_s[4];
    let alive = [1.0, 0.0];
    let dead = [0.0, 1.0];

    let val_0 = value[0], val_1 = value[1];
    let sqrt2_1 = Math.SQRT2 + 1;

    let out_0, out_1;

    if (a <= 1) {
        out_0 = dead[0]; out_1 = dead[1];
    } else if (a > 1 && a <= 2) {
        let f1 = sqrt2_1 * 2 - sqrt2_1 * a;
        let f2 = a - 1;
        out_0 = f1 * dead[0] + f2 * val_0;
        out_1 = f1 * dead[1] + f2 * val_1;
    } else if (a > 2 && a <= 3) {
        let f1 = sqrt2_1 * 3 - sqrt2_1 * a;
        let f2 = a - 2;
        out_0 = f1 * val_0 + f2 * alive[0];
        out_1 = f1 * val_1 + f2 * alive[1];
    } else if (a > 3 && a < 4) {
        let f1 = sqrt2_1 * 4 - sqrt2_1 * a;
        let f2 = a - 3;
        out_0 = f1 * alive[0] + f2 * dead[0];
        out_1 = f1 * alive[1] + f2 * dead[1];
    } else {
        out_0 = dead[0]; out_1 = dead[1];
    }

    let norm = Math.sqrt(out_0 * out_0 + out_1 * out_1);
    if (norm > 0) { out_0 /= norm; out_1 /= norm; }
    return [out_0, out_1];
}

function game_of_life(nhood) {
    let qc = new QuantumCircuit(9);
    let col = 0;

    for (let i = 0; i < 9; i++) {
        let alpha_re = nhood[i][0].re;
        let beta_re = nhood[i][1].re;
        let beta_im = nhood[i][1].im;

        let theta = 2 * Math.acos(Math.max(0, Math.min(1, Math.abs(alpha_re))));
        let phi = 0;
        if (Math.abs(beta_re) > 1e-9 || Math.abs(beta_im) > 1e-9) {
            phi = Math.atan2(beta_im, beta_re);
        }

        qc.addGate("ry", col, i, { params: { theta: theta, phi: theta, lambda: theta } });
        qc.addGate("rz", col + 1, i, { params: { phi: phi, theta: phi, lambda: phi } });
    }
    col += 2;

    for (let q = 0; q < 9; q++) {
        if (q !== 4) {
            let a1 = Math.PI / 2;
            qc.addGate("crx", col++, [4, q], { params: { theta: a1, phi: a1, lambda: a1 } });
        }
        let a2 = 2 * Math.PI / (q + 1);
        qc.addGate("cry", col++, [q, (q + 1) % 9], { params: { theta: a2, phi: a2, lambda: a2 } });
        let a3 = 2 * Math.PI / (((q + 4) % 9) + 1);
        qc.addGate("crz", col++, [q, (q + 1) % 9], { params: { theta: a3, phi: a3, lambda: a3 } });
        if (q !== 4) {
            qc.addGate("cx", col++, [q, 4]);
        }
    }

    qc.run();
    let state = qc.state;

    let rho00 = 0, rho11 = 0, rho01_re = 0, rho01_im = 0;

    let psi_re = new Float64Array(512);
    let psi_im = new Float64Array(512);
    for (let k in state) {
        psi_re[k] = state[k].re;
        psi_im[k] = state[k].im;
    }

    for (let k = 0; k < 512; k++) {
        let p2 = psi_re[k] * psi_re[k] + psi_im[k] * psi_im[k];
        if ((k & 16) === 0) {
            rho00 += p2;
            let k_partner = k | 16;
            rho01_re += psi_re[k] * psi_re[k_partner] + psi_im[k] * psi_im[k_partner];
            rho01_im += psi_im[k] * psi_re[k_partner] - psi_re[k] * psi_im[k_partner];
        } else {
            rho11 += p2;
        }
    }

    return find_closest_pure_state(rho00, rho11, rho01_re, rho01_im);
}

export async function run(params, reportProgress) {
    console.log("[GoL] Entering GoL.run()");
    let image = params.stroke_input.image_rgba;
    let transparentLayer = {
        width: image.width,
        height: image.height,
        data: new Uint8ClampedArray(image.width * image.height * 4),
    };
    let map = MAPPING_TYPES[params.user_input.Mapping];
    let path = params.stroke_input.path;
    let radius = params.user_input.Radius;
    let iterations = params.user_input.Iterations;

    let height = image.height;
    let width = image.width;

    let copy_region;
    if (radius > 0) {
        copy_region = utils.pointsWithinRadius(path, radius, [height, width]);
    } else {
        copy_region = utils.pointsWithinLasso(path, [height, width]);
    }
    console.log(`[GoL] Extracted valid region. Total pixels to process per iteration: ${copy_region.length}. Total iterations: ${iterations}`);

    const extract_neighbourhoods = (y, x) => {
        let neighbours = [];
        for (let dx of [-1, 0, 1]) {
            for (let dy of [-1, 0, 1]) {
                let ny = (y + dy) % height; if (ny < 0) ny += height;
                let nx = (x + dx) % width; if (nx < 0) nx += width;
                neighbours.push([ny, nx]);
            }
        }
        return neighbours;
    };

    let image_hsv = new Float32Array(width * height * 3);
    for (let i = 0; i < width * height; i++) {
        let r = image.data[i * 4] / 255, g = image.data[i * 4 + 1] / 255, b = image.data[i * 4 + 2] / 255;
        let [h, s, v] = colors.rgbToHsv(r, g, b);
        image_hsv[i * 3] = h; image_hsv[i * 3 + 1] = s; image_hsv[i * 3 + 2] = v;
    }

    for (let it = 0; it < iterations; it++) {
        console.log(`[GoL] Starting Iteration ${it + 1} of ${iterations}`);
        let statevector_selection = new Map();
        let semiclassical = new Map();
        let after_iteration = new Map();

        console.time(`GoL-Iter-${it + 1}`);
        for (let i = 0; i < copy_region.length; i++) {
            if (i % 25 === 0 && i > 0) {
                console.log(`[GoL] Iteration ${it + 1}: Processed ${i} / ${copy_region.length} pixels...`);
                if (reportProgress) {
                    reportProgress((it + (i / copy_region.length)) / iterations);
                }
            }

            let y = copy_region[i][0], x = copy_region[i][1];
            let neighbourhood = extract_neighbourhoods(y, x);

            let nhood = [];
            let nhood_s = [];

            for (let n = 0; n < 9; n++) {
                let ny = neighbourhood[n][0], nx = neighbourhood[n][1];
                let key = `${ny},${nx}`;
                if (!statevector_selection.has(key)) {
                    let idx = (ny * width + nx) * 3;
                    let hsv_arr = [image_hsv[idx], image_hsv[idx + 1], image_hsv[idx + 2]];
                    let sv = hsv_to_statevector(hsv_arr, map);
                    statevector_selection.set(key, [sv[0], sv[1]]);
                    semiclassical.set(key, sv[2]);
                }
                nhood.push(statevector_selection.get(key));
                let sc = semiclassical.get(key);
                nhood_s.push([sc, 1 - sc]);
            }

            let [v, purity] = game_of_life(nhood);
            let new_sc = SCGOL(nhood_s)[0];
            let new_hsv = statevector_to_hsv(v, new_sc, map);
            after_iteration.set(`${y},${x}`, new_hsv);
        }

        for (let i = 0; i < copy_region.length; i++) {
            let y = copy_region[i][0], x = copy_region[i][1];
            let new_hsv = after_iteration.get(`${y},${x}`);
            let idx = (y * width + x) * 3;
            image_hsv[idx] = new_hsv[0]; image_hsv[idx + 1] = new_hsv[1]; image_hsv[idx + 2] = new_hsv[2];
        }

        console.timeEnd(`GoL-Iter-${it + 1}`);
        if (reportProgress) reportProgress((it + 1) / iterations);
    }

    for (let i = 0; i < copy_region.length; i++) {
        let y = copy_region[i][0], x = copy_region[i][1];
        let idx = (y * width + x) * 3;
        let [r, g, b] = colors.hsvToRgb(image_hsv[idx], image_hsv[idx + 1], image_hsv[idx + 2]);
        let data_idx = (y * width + x) * 4;
        transparentLayer.data[data_idx] = r * 255;
        transparentLayer.data[data_idx + 1] = g * 255;
        transparentLayer.data[data_idx + 2] = b * 255;
        transparentLayer.data[data_idx + 3] = 255;
    }
    return transparentLayer;
}

File: effect/GoL/GoL.py
import numpy as np
from qiskit import QuantumCircuit, generate_preset_pass_manager
from qiskit_aer import Aer
from qiskit.quantum_info import Pauli, SparsePauliOp, Statevector,partial_trace
from qiskit.circuit.library import RXGate, RZGate,XGate,ZGate,IGate,StatePreparation
import importlib.util
from qiskit import QuantumCircuit, ClassicalRegister, QuantumRegister
from qiskit_aer import Aer, statevector_simulator
from qiskit.quantum_info import partial_trace
import colorsys

import sys
import importlib.util
from pathlib import Path

if getattr(sys, 'frozen', False):
    app_path = Path(sys.executable).parent.parent
else:
    app_path = Path(__file__).resolve().parent.parent.parent

utils_path = app_path / 'effect' / 'utils.py'

spec = importlib.util.spec_from_file_location('utils', str(utils_path))
utils = importlib.util.module_from_spec(spec)
spec.loader.exec_module(utils)
#%%
mapping = ['hsv','hvs','shv','svh','vhs','vsh']

def hsv_to_statevector(hsv,mapping = 'hsv'):
    """
    Maps an RGB value to a qubit statevector and saturation.

    Args:
        rgb: A tuple of (r, g, b) values, where each is in the range [0, 255].

    Returns:
        A tuple of (statevector, saturation), where statevector is a
        numpy array representing the qubit state, and saturation is a float.
    """
    h, s, v = hsv
    hsv = {'h':h,'s':s,'v':v}
    # Map Hue to phase (phi) and Value to amplitude (theta)
    phi = hsv[mapping[0]] * 2 * np.pi
    theta = hsv[mapping[1]] * np.pi

    # Create the statevector [alpha, beta]
    alpha = np.cos(theta / 2)
    beta = np.exp(1j * phi) * np.sin(theta / 2)
    
    statevector = np.array([alpha, beta])
    return statevector, hsv[mapping[2]]

def statevector_to_hsv(statevector, semiclassical, mapping = 'hvs'):
    """
    Maps a qubit statevector and saturation back to an RGB value.

    Args:
        statevector: A numpy array representing the qubit state.
        saturation: The saturation value (as a float).

    Returns:
        A tuple of (r, g, b) values in the range [0, 255].
    """
    
    alpha, beta = statevector[0], statevector[1]

    # Inverse mapping from statevector to H and V
    # Ensure alpha is real and non-negative for acos to be in [0, pi]
    theta = 2 * np.arccos(np.abs(alpha))
    # Handle the case where beta is zero to avoid division by zero
    if np.abs(beta) > 1e-9:
        phi = np.angle(beta) - np.angle(alpha)
    else:
        phi = 0
    
    h = phi / (2 * np.pi)
    # Normalize h to be in [0, 1]
    if h < 0:
        h += 1
    v = theta / np.pi
    return {mapping[0]:h,mapping[1]:v,mapping[2]:semiclassical}
#%%
def find_closest_pure_state(density_matrix):
    """
    Finds the closest pure state to a given density matrix for a single qubit.

    The closest pure state is the eigenvector corresponding to the largest
    eigenvalue of the density matrix.

    Args:
        density_matrix: A 2x2 NumPy array representing the density matrix.
                        It must be Hermitian with a trace of 1.

    Returns:
        A tuple containing:
        - closest_pure_state_vector (np.ndarray): The state vector |ψ⟩.
        - closest_pure_state_density_matrix (np.ndarray): The density matrix |ψ⟩⟨ψ|.

    Raises:
        ValueError: If the input is not a valid 2x2 density matrix.
    """
    # --- Input Validation ---
    if not isinstance(density_matrix, np.ndarray) or density_matrix.shape != (2, 2):
        raise ValueError("Input must be a 2x2 NumPy array.")
    if not np.isclose(np.trace(density_matrix), 1):
        raise ValueError("The trace of the density matrix must be 1.")
    if not np.allclose(density_matrix, density_matrix.conj().T):
        raise ValueError("The density matrix must be Hermitian.")

    # --- Core Algorithm ---
    # For a Hermitian matrix, eigh is preferred as it's more efficient
    # and guarantees real eigenvalues and orthonormal eigenvectors.
    eigenvalues, eigenvectors = np.linalg.eigh(density_matrix)
    purity = np.matmul(density_matrix, density_matrix).trace().real
    # Find the index of the largest eigenvalue
    largest_eigenvalue_index = np.argmax(eigenvalues)

    # The corresponding eigenvector is the state vector of the closest pure state
    closest_pure_state_vector = eigenvectors[:, largest_eigenvalue_index]


    return closest_pure_state_vector, purity

def liveliness(nhood):
    v=nhood
    a = v[0][0]+v[1][0]+v[2][0]+v[3][0]+v[5][0]+v[6][0]+v[7][0]+v[8][0]
    return np.abs(a)
def SCGOL(nhood):
    a = liveliness(nhood)
    value =  nhood[4]
    alive = np.array([1.0,0.0])
    dead = np.array([0.0,1.0])
    B = np.array([[0,0],[1,1]])
    D = np.array([[1,1],[0,0]])
    S = np.array([[1,0],[0,1]])
    if a <= 1:
        value =  dead
    elif (a > 1 and a <= 2):
        value = ((np.sqrt(2)+1)*2-(np.sqrt(2)+1)*a)*dead+(a-1)*value#(((np.sqrt(2)+1)*(2-a))**2+(a-1)**2)
    elif (a > 2 and a <= 3):
        value = (((np.sqrt(2)+1)*3)-(np.sqrt(2)+1)*a)*value+(a-2)*alive#(((np.sqrt(2)+1)*(3-a))**2+(a-2)**2)
    elif (a > 3 and a < 4):
        value = ((np.sqrt(2)+1)*4-(np.sqrt(2)+1)*a)*alive+(a-3)*dead#(((np.sqrt(2)+1)*(4-a))**2+(a-3)**2)
    elif a >= 4:
        value = dead
    value = value/np.linalg.norm(value)
    return value

from qiskit.quantum_info import Operator, Statevector

# Precompute the unitary matrix for the Conway's Game of Life circuit
# This drastically speeds up execution instead of simulating it per pixel.
_gol_qc = QuantumCircuit(9)
for q in range(9):
    if q != 4:
        _gol_qc.crx(np.pi/2, 4, q)
    _gol_qc.cry(2*np.pi/(q+1), q, (q+1)%9)
    _gol_qc.crz(2*np.pi/((q+4)%9+1), q, (q+1)%9)
    if q != 4:
        _gol_qc.cx(q, 4)
GOL_UNITARY = Operator(_gol_qc).data

def game_of_life(nhood):
    v = np.array(nhood).reshape(9,2)
    # Ensure all inputs are normalized
    norms = np.linalg.norm(v, axis=1, keepdims=True)
    v = np.where(norms > 0, v / norms, v)
    
    # Build the 9-qubit statevector
    # Qiskit uses little-endian, so q_n \otimes ... \otimes q_0
    sv = v[0]
    for i in range(1, 9):
        sv = np.kron(v[i], sv)
        
    # Apply the unitary operator
    sv_out = GOL_UNITARY @ sv
    
    # Trace out all qubits except qubit 4
    sv_obj = Statevector(sv_out)
    rho = partial_trace(sv_obj, [0,1,2,3,5,6,7,8])
    
    value, purity = find_closest_pure_state(rho.data)
    return value, purity

def run(params):
    """
    Executes the effect pipeline based on the provided parameters.

    Args:
        parameters (dict): A dictionary containing all the relevant data.

    Returns:
        Image: the new numpy array of RGBA values or None if the effect failed
    """
    
    # Extract image to work from
    image = params["stroke_input"]["image_rgba"]
    map = mapping[params["user_input"]["Mapping"]]
    # It's a good practice to check any of the request variables
    assert image.shape[-1] == 4, "Image must be RGBA format"

    height = image.shape[0]
    width = image.shape[1]
    # Convert the image to HSV colorspace as a copy
    rgb_norm = image[:, :, :3].astype(np.float32) / 255.0

    image_hsv = np.apply_along_axis(
        lambda c: colorsys.rgb_to_hsv(c[0], c[1], c[2]),
        axis=2,
        arr=rgb_norm
    )
    
    # Extract the lasso path
    path = params["stroke_input"]["path"]
    
    radius = params["user_input"]["Radius"]
    if radius > 0:
        copy_region = utils.points_within_radius(path, radius, border = (height, width))
    else:
        copy_region = utils.points_within_lasso(path, border = (height, width))
    def extract_neighbourhoods(x, y):
        neighbours = []
        for dx in [-1, 0, 1]:
            for dy in [-1, 0, 1]:
                ni, nj = (x + dx) % width, (y + dy) % height  # Wrap around edges
                neighbours.append((ni, nj))
        return neighbours
    
    
    for iterations in range(params["user_input"]["Iterations"]):
        copy_selection = {} 
        statevector_selection = {}
        semiclassical = {}
        after_iteration = {}
        for i in copy_region:
            neighbourhood = extract_neighbourhoods(i[0],i[1])
            nhood = np.zeros((9,2), dtype=np.complex128)
            nhood_s = np.zeros((9,2), dtype=np.float64)
            for n,coord in enumerate(neighbourhood):
                if coord not in copy_selection:
                    copy_selection[coord] = np.array(image_hsv[coord[0], coord[1],:3])
                    statevector_selection[coord],semiclassical[coord] = hsv_to_statevector(copy_selection[coord],mapping=map)
                nhood[n] = statevector_selection[coord]
                nhood_s[n] = semiclassical[coord], 1-semiclassical[coord]
            v, purity = game_of_life(nhood)
            new_sc = SCGOL(nhood_s)[0]
            hsv = statevector_to_hsv(v,semiclassical=new_sc, mapping=map)
            after_iteration[(i[0],i[1])] = np.array([hsv['h'],hsv['s'],hsv['v']])
        for key in after_iteration:
            image_hsv[key[0], key[1],:3] = after_iteration[key]
    for key in after_iteration:
        image[key[0], key[1],:3] = np.round(np.array(colorsys.hsv_to_rgb(*after_iteration[key]))*255).astype(np.uint8)
    return image
#%%